### PR TITLE
[SPARK-39451][SQL] Support casting intervals to integrals in ANSI mode

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to support casting of interval types to the integral type: `TINYINT`, `SMALLINT`, `INT`, `BIGINT`.

### Why are the changes needed?
To conform the SQL standard which allows such casting:
Screenshot 2022-06-12 at 13 04 44

### Does this PR introduce _any_ user-facing change?
No, it extends existing behavior.

### How was this patch tested?
By running new tests:
```
$ build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite -- -z cast.sql"
```

### CODE REVIEW

**Comment by MaxGekk** [comment]
cc @srielau

**Comment by srielau** [comment]
What about exact numeric for fractional seconds?
We keep coming back to that….
CAST(INTERVAL ‚12345.678‘ SECOND As DECIMAL(8,3))

**Comment by MaxGekk** [comment]
> What about exact numeric for fractional seconds?

I will implement this separately.  This PR leverages existing functionality of non-ANSI mode.

**Comment by HyukjinKwon** [comment]
cc @gengliangwang FYI

File: sql/core/src/test/resources/sql-tests/inputs/cast.sql [hunk 1]
```diff
@@ -104,3 +104,15 @@ select cast('a' as timestamp_ntz);
 
 select cast(cast('inf' as double) as timestamp);
 select cast(cast('inf' as float) as timestamp);
+
+-- cast ANSI intervals to numerics
+select cast(interval '1' year as tinyint);
+select cast(interval '-10-2' year to month as smallint);
+select cast(interval '1000' month as int);
+select cast(interval -'10.123456' second as tinyint);
+select cast(interval '23:59:59' hour to second as smallint)
```
Comment (gengliangwang) [inline sql/core/src/test/resources/sql-tests/inputs/cast.sql:113]: add `;` at the end of line?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 2]
```diff
@@ -108,8 +108,9 @@ object Cast {
     case (TimestampType, TimestampNTZType) => true
 
     case (StringType, _: CalendarIntervalType) => true
-    case (StringType, _: DayTimeIntervalType) => true
-    case (StringType, _: YearMonthIntervalType) => true
+    case (StringType, _: AnsiIntervalType) => true
+
+    case (_: AnsiIntervalType, _: IntegralType) => true
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:113]: Shall we allow `IntegralType => AnsiIntervalType` as well?

File: sql/core/src/test/resources/sql-tests/inputs/cast.sql (hunk 1 above)
Comment (MaxGekk) [reply sql/core/src/test/resources/sql-tests/inputs/cast.sql:113]: Thanks. I'll add.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala (hunk 2 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:113]: The SQL standard allows this. I think it makes sense to support this too.

**Review by gengliangwang — APPROVED** [review]

**Comment by MaxGekk** [comment]
Merging to master. Thank you, @gengliangwang and @srielau for review.

**Pull request closed without merging**